File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefix = [1] * n
        prefix[0] = 1
        product = 1
        for i in range(1, n):
            prefix[i] = nums[i-1] * prefix[i-1]
        for i in range(n-2, -1, -1):
            product = product * nums[i+1]
            prefix[i] = prefix[i] * product
        return prefix

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int length = nums.length;

        int[] prefix = new int[length];
        prefix[0] = 1;

        for (int i = 1; i < length; i++){
            prefix[i] = prefix[i-1] * nums[i-1];
        }
        int temp = 1;
        for (int i = length-2; i >= 0; i --){
            temp = temp * nums[i+1];
            prefix[i] = prefix[i] * temp;
        }

        return prefix;

    }
}